accepted/2022/version-selection.md

```diff
@@ -0,0 +1,449 @@
+
+Summary of significant changes:
+
+- Disable multi-level lookup.
+- Change default for `global.json` SDK version to a floor version, as opposed to the current pinning behavior.
+- Enable controlling SDK version by environment variable (ENV).
+- Enable proxying to a local/private SDK via the `PATH` SDK.
+- Enable specifying a floor runtime version for an environment.
+
+Note: The last two of those changes are more speculative (less likely to happen).
+
+Let's start by considering some concrete customer scenarios to improve (some of which are overlapping).
+
+## Dev inner loop consistency with lab builds
+
+For many teams, infra builds are the center of their environment. Infra is intended to mean any of Continuous Integration (CI), Pull Request (PR), or Official builds. Teams typically put the most design and implementation effort into their infra builds, with the goal of establishing their most correct and reliable environment. This includes defining their SDK and runtime platform dependencies. In other words, if there is a bug that repros in CI but not on your local dev box, you still have a bug, while the opposite might not be true.
+
+Most devs would prefer a local environment that matches their lab environment. The practice of working in matching environments (local vs lab and across teammates) increases productivity since it inherently avoids spurious local issues or surprise problems in PR builds. There is no good guidance or productized UX that would help teams establish matching environments.
+
+[Ideally, devs would clone a repo](https://github.com/dotnet/sdk/issues/8254), `dotnet build` a solution file or double click on it (to be opened in Visual Studio) and the rest of establishing a matching environment would be taken care of. We're a ways off that. Let's consider what folks do today.
+
+Many repos (including .NET repos) follow a pattern of running a script (often run as part of the build) that downloads a specific patch of the .NET SDK, installs it into a private directory (often just off the root of the repo) and then runs the build and tests for the repo. This process has a lot of benefits, primarily that everyone working on the repo (at least on the same branch) uses the same SDK version (whether they have the SDK installed or not). That patterns also makes it very quick and convenient to build a repo on a new machine.
```

Comment:
For many of our repos, there's even an additional script (sdk-build-env.bat in the sdk repo) that developers run to configure using the local dotnet hive so we can use that with VS. Additionally, there's a third script (dogfood.cmd) that lets us use the bits that we just built (though that may be more specific to .NET).

Reply:
Interesting. I was unaware of that. It strengthens the same point that we have a well-defined convention for our repos.